### PR TITLE
Add Python 3.13 (beta) support

### LINKED ISSUE #12323: Test failures on Python 3.13.0b1
I don't seem to be able to find a bug for tracking Python 3.13 support, so I hope it's fine if I file one.

```pytb
$ tox -e py313
.pkg: _optional_hooks> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_editable> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /usr/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py313: install_package> python -I -m pip install --force-reinstall --no-deps /tmp/pytest/.tox/.tmp/package/2/pytest-8.3.0.dev38+g93dd34e76.tar.gz
py313: commands[0]> pytest
============================= test session starts ==============================
platform linux -- Python 3.13.0b1, pytest-8.3.0.dev38+g93dd34e76, pluggy-1.5.0
cachedir: .tox/py313/.pytest_cache
rootdir: /tmp/pytest
configfile: pyproject.toml
testpaths: testing
plugins: hypothesis-6.102.1
collected 3668 items

[…]
=========================== short test summary info ============================
FAILED testing/code/test_excinfo.py::TestTraceback_f_g_h::test_traceback_recursion_index
FAILED testing/code/test_excinfo.py::TestFormattedExcinfo::test_repr_traceback_recursion
FAILED testing/code/test_excinfo.py::test_exception_repr_extraction_error_on_recursion
FAILED testing/code/test_source.py::test_getfslineno - assert 368 == -1
FAILED testing/test_collection.py::TestSession::test_collect_protocol_single_function
FAILED testing/test_collection.py::TestSession::test_collect_custom_nodes_multi_id
FAILED testing/test_collection.py::TestSession::test_collect_subdir_event_ordering
FAILED testing/test_collection.py::TestSession::test_collect_two_commandline_args
FAILED testing/test_doctest.py::TestDoctests::test_doctest_unexpected_exception
FAILED testing/test_doctest.py::TestDoctests::test_doctest_linedata_on_property
FAILED testing/test_legacypath.py::test_testdir_makefile_ext_none_raises_type_error
XPASS testing/_py/test_local.py::TestLocalPath::test_make_numbered_dir_multiprocess_safe - #11603
= 11 failed, 3535 passed, 110 skipped, 11 xfailed, 1 xpassed in 138.50s (0:02:18) =
py313: exit 1 (139.10 seconds) /tmp/pytest> pytest pid=65870
  py313: FAIL code 1 (145.84=setup[6.73]+cmd[139.10] seconds)
  evaluation failed :( (146.00 seconds)
```

This is on Gentoo Linux amd64, with 93dd34e76d9c687d1c249fe8cf94bdf46813f783.

Full log: 
[py313.txt](https://github.com/pytest-dev/pytest/files/15310368/py313.txt)

### PR DESCRIPTION
**Pull request by bluetech**
Fix #12323

### CODE REVIEW

File: src/_pytest/pytester.py [hunk 1]
```diff
@@ -289,7 +289,8 @@ def assert_contains(self, entries: Sequence[Tuple[str, str]]) -> None:
         __tracebackhide__ = True
         i = 0
         entries = list(entries)
-        backlocals = sys._getframe(1).f_locals
+        # Since Python 3.13, f_locals is not a dict, but eval requires a dict.
```
Comment (bluetech) [inline src/_pytest/pytester.py:292]: https://discuss.python.org/t/pep-667-consistent-views-of-namespaces/46631/36

File: src/_pytest/pytester.py [hunk 2]
```diff
@@ -760,6 +761,9 @@ def _makefile(
     ) -> Path:
         items = list(files.items())
 
+        if ext is None:
```
Comment (bluetech) [inline src/_pytest/pytester.py:764]: https://github.com/python/cpython/issues/119113

**Comment by bluetech** [comment]
One remaining failure, on windows:

```
 ________________________ test_cache_reportheader[env1] ________________________

env = ('TOX_ENV_DIR', '/tox_env_dir')
pytester = <Pytester WindowsPath('C:/Users/runneradmin/AppData/Local/Temp/pytest-of-unknown/pytest-0/test_cache_reportheader1')>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x00000209B926A120>

    @pytest.mark.parametrize("env", ((), ("TOX_ENV_DIR", "/tox_env_dir")))
    def test_cache_reportheader(
        env: Sequence[str], pytester: Pytester, monkeypatch: MonkeyPatch
    ) -> None:
        pytester.makepyfile("""def test_foo(): pass""")
        if env:
            monkeypatch.setenv(*env)
            expected = os.path.join(env[1], ".pytest_cache")
        else:
            monkeypatch.delenv("TOX_ENV_DIR", raising=False)
            expected = ".pytest_cache"
        result = pytester.runpytest("-v")
>       result.stdout.fnmatch_lines([f"cachedir: {expected}"])
E       Failed: nomatch: 'cachedir: /tox_env_dir\\.pytest_cache'
E           and: '============================= test session starts ============================='
E           and: 'platform win32 -- Python 3.13.0b1, pytest-8.3.0.dev49+g87c7a72ce, pluggy-1.5.0 -- D:\\a\\pytest\\pytest\\.tox\\py313\\Scripts\\python.EXE'
E           and: 'cachedir: C:\\tox_env_dir\\.pytest_cache'
E           and: 'rootdir: C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pytest-of-unknown\\pytest-0\\test_cache_reportheader1'
E           and: 'collecting ... collected 1 item'
E           and: ''
E           and: 'test_cache_reportheader.py::test_foo PASSED                              [100%]'
E           and: ''
E           and: '============================== 1 passed in 0.01s =============================='
E       remains unmatched: 'cachedir: /tox_env_dir\\.pytest_cache'
```

Probably something in os.path/pathlib changed.

**Comment by RonnyPfannschmidt** [comment]
Seems like its normalized/resolved now, so the env name is no longer correct

**Comment by bluetech** [comment]
I don't know if the `C:\` is correct or not, and I don't have windows handy, but since it's only for display, and probably tox doesn't issue such env dir path on windows anyway, so decided to sidestep the issue in the test.

This is ready for review, I'll also backport it for 8.2.1.

File: src/_pytest/pytester.py (hunk 2 above)
Comment (nicoddemus) [inline src/_pytest/pytester.py:764]: Let's leave a note here:

```suggestion
        # https://github.com/python/cpython/issues/119113  
        if ext is None:
```

File: changelog/12334.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Support for Python 3.13 (beta1 at the time of writing).
```
Comment (nicoddemus) [inline changelog/12334.bugfix.rst:1]: I would categorize this as a feature actually.

File: src/_pytest/pytester.py (hunk 2 above)
Comment (bluetech) [reply src/_pytest/pytester.py:764]: I don't think it's necessary, since I think we better not rely on it and just keep the check.

Comment (nicoddemus) [reply src/_pytest/pytester.py:764]: I don't see what harm a small comment like that could cause either, but OK.

Comment (nicoddemus) [reply src/_pytest/pytester.py:764]: Also in that case, we probably should change the signature as well, so we have the typing match the runtime behavior.

Comment (bluetech) [reply src/_pytest/pytester.py:764]: The type is `ext: str` which is correct. The callers have some `*args, **kwargs` forwarding which is somewhat annoying to type.

Comment (nicoddemus) [reply src/_pytest/pytester.py:764]: Well OK, this is no big deal anyway.

**Pull request merged**